### ISSUE TITLE
Incorrect compile error when having a 'const' inside a function expression of same name

### ISSUE THREAD

**Issue by mor-n4** [babel/babel#6302]
This is a bug report.

Babel shows a compile-time error for the below snippet, even though it is valid ES2015.

### Input Code

[REPL](https://babeljs.io/repl/#?babili=false&evaluate=true&lineWrap=false&presets=es2015&targets=&browsers=&builtIns=false&debug=false&code=(function%20fn()%20%7B%0A%20%20const%20fn%20%3D%2042%3B%0A%7D)%3B)

```js
(function fn() {
  const fn = 42;
});
```

### Babel Configuration (.babelrc, package.json, cli command)

```js
{
  "presets": ["es2015"]
}
```

### Expected Behavior

No compile error should occur, as the above code is valid ES2015 (tested with node.js).

### Current Behavior

The following incorrect compile error is reported:
```
repl: "fn" is read-only
> 1 | (function fn() {
    |  ^
  2 |   const fn = 42;
  3 | });
```
Interestingly, the error disappears if either the 'const' is renamed, the function expression is renamed, or the outer parentheses are removed (i.e. function expression is turned into a declared function). So, to reproduce the bug, the function and 'const' must have the same name and the function has to be a function expression (not a function declaration).

### Possible Solution

n/a

### Context

n/a

### Your Environment

| software         | version(s)
| ---------------- | -------
| Babel            |  v6.26.0
| Operating System | Mac OS X

**Comment by wcastand** [comment]
Hi, can someone help write a test case for this bug. I tried to add in babel-traverse tests a fixture with the code. (actual.js, expected.js, option.json) but the tests are still passing.

I thought this was going to failed since it's supposed to throw an error.
I tried to make a typo in the test's code and the test throw and error so they are taken into account.

Maybe i don't have the right approach ?
